# Анализ базы данных сервиса для чтения книг

In [1]:
!pip install SQLAlchemy
!pip install psycopg2 

In [2]:
%run passwords.ipynb

In [3]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': PWD, # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## 1. Проверка данных

In [4]:
def preprocess(df):
    """Принимает на вход датасет, возвращает первые 5 строк, общую информацию, доли пропусков и число дубликатов"""
    display(df.head())
    display(df.info())
    display(round((df.isna().mean()*100),2))
    display(df.duplicated().sum())

In [5]:
b = ''' SELECT *
        FROM books;
    '''
books = pd.io.sql.read_sql(b, con = engine) 
preprocess(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id             0.0
author_id           0.0
title               0.0
num_pages           0.0
publication_date    0.0
publisher_id        0.0
dtype: float64

0

***
Датасет `books` содержит 1000 строк, пропусков и дубликатов нет. 

In [6]:
a = ''' SELECT *
        FROM authors;
    '''
authors = pd.io.sql.read_sql(a, con = engine) 
preprocess(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id    0.0
author       0.0
dtype: float64

0

***
Датасет `authors` содержит 636 строк, пропусков и дубликатов нет. 

In [7]:
p = ''' SELECT *
        FROM publishers;
    '''
publishers = pd.io.sql.read_sql(p, con = engine) 
preprocess(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id    0.0
publisher       0.0
dtype: float64

0

***
Датасет `pulishers` содержит 340 строк, пропусков и дубликатов нет. 

In [8]:
r_1 = ''' SELECT *
          FROM ratings;
      '''
ratings = pd.io.sql.read_sql(r_1, con = engine) 
preprocess(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id    0.0
book_id      0.0
username     0.0
rating       0.0
dtype: float64

0

***
Датасет `ratings` содержит 6456 строк, пропусков и дубликатов нет. 

In [9]:
r_2 = ''' SELECT *
          FROM reviews;
      '''
reviews = pd.io.sql.read_sql(r_2, con = engine) 
preprocess(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id    0.0
book_id      0.0
username     0.0
text         0.0
dtype: float64

0

***
Датасет `reviews` содержит 2793 строк, пропусков и дубликатов нет. 

## 2. Расчеты

### 2.1. Количество книг, вышедших после 1 января 2000 года

In [10]:
query_1 = ''' SELECT COUNT(book_id) AS books_cnt
              FROM books
              WHERE publication_date > '2000-01-01';
          '''

In [11]:
books_cnt = pd.io.sql.read_sql(query_1, con = engine) 
books_cnt

books_cnt
0        819

После 1 января 2000 года вышло **819 книг**.

### 2.2. Количество обзоров и средняя оценка для каждой книги 

In [12]:
query_2 = ''' SELECT
                  books.title AS title,
                  SUBQ1.avg_rating AS avg_rating,
                  SUBQ2.reviews_cnt AS reviews_cnt
              FROM
                  books
                  LEFT JOIN
                          (SELECT 
                              book_id,
                              AVG(rating) AS avg_rating
                          FROM 
                              ratings
                          GROUP BY
                              book_id) AS SUBQ1 ON SUBQ1.book_id = books.book_id
                  LEFT JOIN                  
                          (SELECT 
                              book_id,
                          COUNT(review_id) AS reviews_cnt
                          FROM 
                              reviews
                          GROUP BY
                              book_id) AS SUBQ2 ON SUBQ2.book_id = books.book_id;
            '''

In [13]:
books_grouped = pd.io.sql.read_sql(query_2, con = engine) 
books_grouped.head()

title  avg_rating  reviews_cnt
0          The Body in the Library (Miss Marple  #3)    4.500000          2.0
1                                          Galápagos    4.500000          2.0
2                           A Tree Grows in Brooklyn    4.250000          5.0
3  Undaunted Courage: The Pioneering First Missio...    4.000000          2.0
4                                        The Prophet    4.285714          4.0

Проверка суммарного числа обзоров для всех книг (должно быть 2793):

In [14]:
books_grouped['reviews_cnt'].sum()

2793.0

### 2.3. Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [15]:
query_3 = ''' SELECT 
                  publishers.publisher AS publisher,
                  COUNT(books.book_id) AS books_cnt
              FROM 
                  books
                  INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
              WHERE 
                  books.num_pages > 50
              GROUP BY publisher
              ORDER BY books_cnt DESC
              LIMIT 1;
          '''

In [16]:
top_publisher = pd.io.sql.read_sql(query_3, con = engine) 
top_publisher

publisher  books_cnt
0  Penguin Books         42

**Penguin Books** - издательство, выпустившее наибольшее число книг (42) толще 50 страниц.

### 2.4. Автор с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками)

In [17]:
query_4 = ''' SELECT 
                  authors.author AS author,
                  AVG(ratings.rating) AS avg_rating
              FROM 
                  books
                  INNER JOIN authors ON authors.author_id = books.author_id
                  INNER JOIN (
                                   SELECT 
                                       book_id,
                                       COUNT(rating) AS rating_cnt
                                   FROM ratings
                                   GROUP BY book_id
                                   HAVING COUNT(rating_id) >= 50) AS SUBQ ON SUBQ.book_id = books.book_id
                  INNER JOIN ratings ON ratings.book_id=books.book_id
              GROUP BY author
              ORDER BY avg_rating DESC
              LIMIT 1;
          '''

In [18]:
top_author = pd.io.sql.read_sql(query_4, con = engine) 
top_author

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Самую высокую среднюю оценку книг (4.3) среди книг с 50 и более оценками получила **Джоан Роулинг**.

### 2.5. Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
query_5 = ''' SELECT
                  AVG(SUBQ_2.reviews_cnt) AS avg_reviews
              FROM 
                  (SELECT 
                  SUBQ.username AS user,
                  COUNT(reviews.review_id) AS reviews_cnt
              FROM 
                  reviews
                  INNER JOIN (SELECT
                                   username,
                                   COUNT(rating) AS rating_cnt
                              FROM ratings
                              GROUP BY username
                              HAVING COUNT(rating_id) > 50) AS SUBQ ON SUBQ.username = reviews.username
              GROUP BY SUBQ.username) AS SUBQ_2;
          '''

In [20]:
avg_reviews = pd.io.sql.read_sql(query_5, con = engine) 
avg_reviews

avg_reviews
0    24.333333

Пользователи, которые поставили более 50 оценок, в среднем написали по **24.3 обзора**.